feat: Add interfaces for new scanner implementation (#123)
kyma-project/kyma-metrics-collector: docs/contributor/arch/01-collector-structure.md

# 1. Collector Structure

Date: 2024-11-20

## Status

Proposed

## Context

> The current implementation of KMCs SKR processing has been built with a monolithic architecture. This has led to a number of issues including:

Reviewer: ```suggestion
The current implementation of KMC's SKR processing has been built with a monolithic architecture. This has led to a number of issues, including:
```

> - a single issue processing an SKR renders the entire billing process for this cluster invalid

Reviewer: wow that's wild :O

- adding new ressources to the billing process is difficult
- the current implementation is hard to test

> - every client needs to implement 

Reviewer: implement... what?

```suggestion
- every client must implement [what]
```

## Decision

> We will refactor the current implementation of the SKR processing to a more modular architecture. This will allow us to:

Reviewer: ```suggestion
We will refactor the current implementation of the SKR processing to a more modular architecture. This will enable us to:
```

- process each scan result independently
- add new resources to scan easily

> - each scan result will contain its own logic for processing the data. Processing will include:

Reviewer: `This will allow us to each scan will contain its own logic...`
This sentence doesn't make sense.

Neither do the following sentences.
`This will allow us to extracting all requited information...`
`This will allow us to converting the scan result`

How about:
line 19: "This has the following advantages:
- We can process...
- We can add...
- Each new scan result will...
- We can extract... {if it's "we" who do that?}
- We can convert... {if it's "we" who do that?}"

>   - extracting all required information for unified metering (eg. capacity units calculated for the resource, additional non-billable metrics)

Reviewer: ```suggestion
  - extracting all required information for unified metering (such as capacity units calculated for the resource, additional non-billable metrics)
```

  - converting the scan result to storage / cpu / memory units for the EDP backend

> The interfaces and their purpose is defined as follows:

Reviewer: ```suggestion
The interfaces and their purpose are defined as follows:
```

> - `Scanner` is an interface for extracting a specific resource related to a single cluster

Reviewer: ```suggestion
- `Scanner` is an interface for extracting a specific resource related to a single cluster.
```

> - `ScanConverter` is the interface required for all ScanResults. It specifies converting a result to a backend specific measurement

Reviewer: ```suggestion
- `ScanConverter` is the interface required for all ScanResults. It specifies converting a result to a backend-specific measurement.
```

> - `CollectorSender` is an interface for collecting and sending the scan results (ScanConverter interface) to the backend

Reviewer: ```suggestion
- `CollectorSender` is an interface for collecting and sending the scan results (ScanConverter interface) to the backend.
```
